src/grpc/zero_copy_stream_test.cc

```diff
@@ -166,68 +176,206 @@ TEST_F(GrpcZeroCopyInputStreamTest, Backups) {
   // Check the message delimiter
   const void *data = nullptr;
   int size = -1;
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(5, size);
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(5, size);
   EXPECT_EQ(slice1.size() + slice2.size(),
             DelimiterToSize(reinterpret_cast<const unsigned char *>(data)));
 
   // Back up
   stream.BackUp(5);
 
-  // Test the BytesAvailable()
+  // Test BytesAvailable() and ByteCount()
   EXPECT_EQ(slice1.size() + slice2.size() + 5,  // +5 bytes for the delimiter
             stream.BytesAvailable());
+  EXPECT_EQ(0, stream.ByteCount());
 
   // Test the slice again
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(5, size);
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(5, size);
   EXPECT_EQ(slice1.size() + slice2.size(),
             DelimiterToSize(reinterpret_cast<const unsigned char *>(data)));
 
   // Test the first slice
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(slice1.size(), size);
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(slice1.size(), size);
   EXPECT_EQ(slice1, std::string(reinterpret_cast<const char *>(data), size));
 
   // Back up & test again
   stream.BackUp(size);
   EXPECT_EQ(slice1.size() + slice2.size(), stream.BytesAvailable());
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(slice1.size(), size);
+  EXPECT_EQ(5, stream.ByteCount());
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(slice1.size(), size);
   EXPECT_EQ(slice1, std::string(reinterpret_cast<const char *>(data), size));
 
   // Now Back up 10 bytes & test again
   stream.BackUp(10);
   EXPECT_EQ(10 + slice2.size(), stream.BytesAvailable());
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(10, size);
+  EXPECT_EQ(5 + slice1.size() - 10, stream.ByteCount());
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(10, size);
   EXPECT_EQ(slice1.substr(slice1.size() - 10),
             std::string(reinterpret_cast<const char *>(data), size));
 
   // Test the second slice
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(slice2.size(), size);
+  EXPECT_TRUE(stream.Next(&data, &size));
+  EXPECT_EQ(slice2.size(), size);
   EXPECT_EQ(slice2, std::string(reinterpret_cast<const char *>(data), size));
 
   // Back up and test again
   stream.BackUp(size);
   EXPECT_EQ(slice2.size(), stream.BytesAvailable());
-  ASSERT_TRUE(stream.Next(&data, &size));
-  ASSERT_EQ(slice2.size(), size);
+  EXPECT_EQ(5 + slice1.size(), stream.ByteCount());
```

Comment:
Are you sure `ByteCount` should double count the bytes after a back up? I didn't find any documentation on this.

Reply:
ByteCount() should not double count,  it should be the count from start of stream to current position.